pkg/codegen/internal/tstypes/tstypes_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tstypes
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+
+	"github.com/stretchr/testify/assert"
+	"pgregory.net/rapid"
+)
+
+func TestParenInsert(t *testing.T) {
+	ast := astGenerator()
+	rapid.Check(t, func(t *rapid.T) {
+		example := ast.Draw(t, "ast").(TypeAst)
+
+		t.Logf("example: %s", spew.Sdump(example))
+		t.Logf("example type: %s", TypeLiteral(example))
+
+		tokens := (&typeScriptTypeUnparser{}).unparse(example)
+
+		parsed, err := (&typeScriptTypeParser{}).parse(tokens)
+
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		t.Logf("parsed: %s", spew.Sdump(parsed))
+		assert.Equal(t, example, parsed)
+	})
+}
+
+func astGenerator() *rapid.Generator {
```

Comment:
Lib remark: shrinking trees seems not entirely satisfactory, I'd like to compare to what Haskell has and if necessary send a patch upstream. Once a bad tree T has been found, I'd like it to be minimized heavily by pruning branches that do not matter for the test before presenting to the user.